core/base/batch_multi_vector.cpp

```diff
@@ -0,0 +1,256 @@
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+
+
+#include <algorithm>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/base/utils.hpp>
+
+
+#include "core/base/batch_multi_vector_kernels.hpp"
+
+
+namespace gko {
+namespace batch_multi_vector {
+
+
+GKO_REGISTER_OPERATION(scale, batch_multi_vector::scale);
+GKO_REGISTER_OPERATION(add_scaled, batch_multi_vector::add_scaled);
+GKO_REGISTER_OPERATION(compute_dot, batch_multi_vector::compute_dot);
+GKO_REGISTER_OPERATION(compute_norm2, batch_multi_vector::compute_norm2);
+GKO_REGISTER_OPERATION(copy, batch_multi_vector::copy);
+
+
+}  // namespace batch_multi_vector
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::scale_impl(
+    const BatchMultiVector<ValueType>* alpha)
+{
+    GKO_ASSERT_EQ(alpha->get_num_batch_entries(),
+                  this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_ROWS(alpha->get_common_size(), dim<2>(1, 1));
+    for (size_type b = 0; b < alpha->get_num_batch_entries(); ++b) {
+        if (alpha->get_common_size()[1] != 1) {
+            // different alpha for each column
+            GKO_ASSERT_EQUAL_COLS(this->get_common_size(),
+                                  alpha->get_common_size());
+        }
+    }
```

Comment:
I don't think this loop does anything.
```suggestion
      if (alpha->get_common_size()[1] != 1) {
          GKO_ASSERT_EQUAL_COLS(this->get_common_size(),
                                alpha->get_common_size());
      }
```